app/models/vm/operations.rb

```diff
@@ -15,14 +15,28 @@ module Vm::Operations
 
   def cockpit_url
     return if ipaddresses.blank?
-    miq_server = ext_management_system.nil? ? nil : ext_management_system.zone.remote_cockpit_ws_miq_server
+    miq_server = if ext_management_system.nil? || ext_management_system.zone.remote_cockpit_ws_miq_server.nil?
+                   nil
+                 else
+                   ext_management_system.zone.remote_cockpit_ws_miq_server
+                 end
     MiqCockpit::WS.url(miq_server,
-                       MiqCockpitWsWorker.fetch_worker_settings_from_server(miq_server),
-                       ipv4_address || ipaddresses.first)
+                       miq_server.nil? ? nil : MiqCockpitWsWorker.fetch_worker_settings_from_server(miq_server),
+                       ipv4_address || ipaddresses.first).to_s
   end
 
   def ipv4_address
-    ipaddresses.find { |ip| IPAddr.new(ip).ipv4? }
+    if %w(amazon google).include?(vendor.downcase)
+      public_address
+    elsif %w(openstack).include?(vendor.downcase)
```

Comment:
Please no provider specific code in the base module.  This should be handled via provider overrides.